addon/components/bs-modal.js

```diff
@@ -485,81 +515,101 @@ export default class Modal extends Component {
-      this.args.onHidden?.();
-    });
+    await this.hideBackdrop();
+
+    if (this.isDestroyed) {
+      return;
+    }
+
+    this.removeBodyClass();
+    this.resetAdjustments();
+    this.resetScrollbar();
+    this.state = 'closed';
+    this.args.onHidden?.();
   }
 
   /**
-   * SHow/hide the backdrop
+   * Show the backdrop
    *
-   * @method handleBackdrop
+   * @method showBackdrop
    * @param callback
    * @private
    */
-  handleBackdrop(callback) {
-    let doAnimate = this.usesTransition;
-
-    if (this.open && this.backdrop) {
-      this.showBackdrop = true;
-
-      if (!callback) {
-        return;
-      }
+  async showBackdrop() {
+    if (!this.backdrop) {
+      return;
+    }
 
-      next(() => {
-        let backdrop = this.backdropElement;
-        assert('Backdrop element should be in DOM', backdrop);
-        if (doAnimate) {
-          transitionEnd(backdrop, this.backdropTransitionDuration).then(callback);
-        } else {
-          callback();
-        }
-      });
-    } else if (!this.open && this.backdrop) {
-      let backdrop = this.backdropElement;
-      assert('Backdrop element should be in DOM', backdrop);
+    this.isBackdropShown = true;
 
-      let callbackRemove = () => {
-        if (this.isDestroyed) {
+    await new Promise((resolve) => {
+      next(async () => {
+        if (!this.backdrop) {
           return;
         }
-        this.showBackdrop = false;
-        if (callback) {
-          callback.call(this);
+
+        let { backdropElement, usesTransition } = this;
+        assert('Backdrop element should be in DOM', backdropElement);
+
+        if (usesTransition) {
+          await transitionEnd(backdropElement, this.backdropTransitionDuration);
         }
-      };
-      if (doAnimate) {
-        transitionEnd(backdrop, this.backdropTransitionDuration).then(callbackRemove);
-      } else {
-        callbackRemove();
-      }
-    } else if (callback) {
-      next(this, callback);
-    }
+
+        resolve();
+      });
+    });
   }
 
   /**
-   * Attach/Detach resize event listeners
+   * Hide the backdrop
    *
-   * @method resize
+   * @method hideBackdrop
    * @private
    */
-  resize() {
-    if (this.open) {
-      window.addEventListener('resize', this.adjustDialog, false);
-    } else {
-      window.removeEventListener('resize', this.adjustDialog, false);
+  async hideBackdrop() {
+    if (!this.backdrop) {
+      return;
+    }
+
+    let { backdropElement, usesTransition } = this;
+    assert('Backdrop element should be in DOM', backdropElement);
+
+    if (usesTransition) {
+      await transitionEnd(backdropElement, this.backdropTransitionDuration);
+    }
+
+    if (this.isDestroyed) {
+      return;
     }
+
+    this.isBackdropShown = false;
+  }
+
+  /**
+   * Attach resize event listeners
+   */
+  addResizeEventListeners() {
+    if (isFastBoot(this)) {
+      return;
+    }
+
+    window.addEventListener('resize', this.adjustDialog, false);
```

Comment:
we might also use `{{on-window ...}}` from https://github.com/buschtoens/ember-on-helper to attach listener more declaratively in the template!?

Reply:
Great suggestion. Done in 708adc7e97cffbc38d52fec70a4239fd2c5bbce7.